### DISCUSSION TITLE
How to add new platform-independent example?

### DISCUSSION THREAD

**Discussion by AminoACID123** [contiki-ng/contiki-ng#2570] [Q&A]
I want to add a new platform-independent example under the `examples` directory.

**Comment by pjonsson** [comment]
The easiest way is probably to copy `examples/hello-world` and modify to fit your needs.

Here's documentation about various relevant parts of Contiki-NG:
https://docs.contiki-ng.org/en/release-v4.9/doc/getting-started/The-Contiki-NG-build-system.html
https://docs.contiki-ng.org/en/release-v4.9/doc/getting-started/The-Contiki-NG-configuration-system.html
https://docs.contiki-ng.org/en/release-v4.9/doc/getting-started/The-Contiki-NG-logging-system.html

**Reply by AminoACID123** [reply]
The problem is I do not know how to modify the Makefile. The `Makefile` in `examples/hello-world` has the following two lines:
```
CONTIKI_PROJECT = hello-world
all: $(CONTIKI_PROJECT)
```
So target `all` depends on `hello-world`, but I could not find anywhere in the project that specifies the rule for making target `hello-world`.

**Reply by pjonsson** [reply]
That comes from the last lines of Makefile.include:
```Makefile
# Build this target.
# Match-anything pattern rule to allow the project makefiles to
# abstract from the actual binary name. It needs to contain some
# command in order to be a rule, not just a prerequisite.
%: %.$(TARGET)
	@
```

I don't think you want to try to untangle exactly how the build system figures out what to build, but here are the rules for building object files from C files, and linking the final binary:

```Makefile
ifndef CUSTOM_RULE_C_TO_OBJECTDIR_O
$(OBJECTDIR)/%.o: %.c | $(DEPDIR)
	$(TRACE_CC)
	$(Q)$(CCACHE) $(CC) $(CFLAGS) $(DEPFLAGS) -c $< -o $@
endif

ifndef CUSTOM_RULE_LINK
# Targets can define LD_START_GROUP and LD_END_GROUP to resolve circular
# dependencies between linked libraries, see:
# https://stackoverflow.com/questions/5651869/gcc-what-are-the-start-group-and-end-group-command-line-options/5651895
# These are not defined by default since it has a significant performance cost.
TARGET_LIBEXTRAS = $(LD_START_GROUP) $(TARGET_LIBFILES) $(LD_END_GROUP)

# The order of object files passed to $(LD) varies across runs and this causes
# binary sizes to vary. Sort the files to get a deterministic size.
#
# Cooja passes LIBNAME through the environment and all .csc files contain
# calls to make with a target that is *not* what Cooja needs. Add
# a compatibility line for link rule so we can have a single rule that
# outputs to LIBNAME on all platforms.
$(BUILD_DIR_BOARD)/%.$(TARGET): LIBNAME ?= $@
$(BUILD_DIR_BOARD)/%.$(TARGET): $(OBJECTDIR)/%.o $(LDSCRIPT) $(PROJECT_OBJECTFILES) $(PROJECT_LIBRARIES) $(CONTIKI_OBJECTFILES)
	$(TRACE_LD)
	$(Q)$(LD) $(LDFLAGS) $(TARGET_STARTFILES) $(sort ${filter-out $(LDSCRIPT) %.a,$^}) ${filter %.a,$^} $(TARGET_LIBEXTRAS) -o $(LIBNAME)
ifdef BINARY_SIZE_LOGFILE
	$(Q)$(SIZE) $(LIBNAME) | grep $(BUILD_DIR_BOARD) >> $(BINARY_SIZE_LOGFILE)
endif
endif
```
ARM-based platforms still has a custom link rule, but that does the same thing, and the custom rule will be removed with #2483.

**Reply by AminoACID123** [reply]
I see. Thanks for the reply.
